ydb/src/load_balancer/nearest_dc_balancer.rs

```diff
@@ -0,0 +1,388 @@
+use std::{
+    borrow::{Borrow, BorrowMut},
+    collections::HashMap,
+    sync::Arc,
+};
+
+use http::Uri;
+use itertools::Itertools;
+use rand::{seq::SliceRandom, thread_rng};
+use std::time::Duration;
+use tokio::{
+    io::AsyncWriteExt,
+    net::TcpStream,
+    sync::{
+        broadcast, mpsc,
+        watch::{self, Sender},
+        Mutex,
+    },
+    task::JoinSet,
+};
+use tokio_util::sync::CancellationToken;
+use tracing::{error, info, warn};
+
+use crate::{
+    discovery::NodeInfo,
+    grpc_wrapper::raw_services::Service,
+    waiter::{AllWaiter, WaiterImpl},
+    DiscoveryState, Waiter, YdbError, YdbResult,
+};
+
+use super::{random_balancer::RandomLoadBalancer, LoadBalancer};
+
+pub(crate) struct BalancerConfig {
+    pub(super) fallback_strategy: FallbackStrategy,
+}
+
+#[derive(Default)]
+pub(super) struct BalancerState {
+    pub(super) preferred_endpoints: Vec<NodeInfo>,
+}
+
+// What will balancer do if there is no available endpoints at local dc
+pub(crate) enum FallbackStrategy {
+    Error,                                   // Just throw error
+    BalanceWithOther(Box<dyn LoadBalancer>), // Use another balancer
+}
+
+impl Default for BalancerConfig {
+    fn default() -> Self {
+        BalancerConfig {
+            fallback_strategy: FallbackStrategy::BalanceWithOther(Box::new(
+                RandomLoadBalancer::new(),
+            )),
+        }
+    }
+}
+
+pub(crate) struct NearestDCBalancer {
+    discovery_state: Arc<DiscoveryState>,
+    state_sender: Sender<Arc<DiscoveryState>>,
+    ping_token: CancellationToken,
+    waiter: Arc<WaiterImpl>,
+    config: BalancerConfig,
+    balancer_state: Arc<Mutex<BalancerState>>,
```

Comment:
what about use RWLock instead of Mutex?
it will allow multiply reads in parallel and block for update state only

Reply:
Because there is only reader and only one writer